### PR TITLE
feat: notify hmcts of placement application upload (FPLA-821)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-821

### Change description ###
Uploading a placement application will now send a notification provided the newly uploading placement is not a re-upload of the previous placement. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 1]
```diff
@@ -536,6 +542,34 @@ void shouldNotifyLocalAuthorityOfIssuedStandardDirectionsOrder() throws IOExcept
             eq("12345"));
     }
 
+    @Test
+    void shouldNotifyAdminOfPlacementApplicationUpload() throws Exception {
+        CallbackRequest callbackRequest = callbackRequest();
+        CaseDetails caseDetails = callbackRequest().getCaseDetails();
+        final Map<String, Object> expectedParameters = getExpectedPlacementNotificationParameters();
+
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: can remove one of these whitespace lines

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java [hunk 2]
```diff
@@ -45,11 +63,15 @@ void shouldUpdateChildPlacement() {
...
 
         CaseDetails caseDetails = CaseDetails.builder()
-            .data(Map.of(
-                "children1", List.of(child1, child2),
-                "placements", List.of(child2Placement, child1Placement),
-                "placement", child2NewPlacement.getValue(),
-                "childrenList", child2.getId()))
+            .id(CASE_ID)
+            .data(ImmutableMap.<String, Object>builder()
+                .putAll(buildNotificationData())
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: Could build this in a private method?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java [hunk 3]
```diff
@@ -64,6 +86,31 @@ void shouldUpdateChildPlacement() {
...
+    }
+
+    @Test
+    void shouldNotSendANotificationWhenUploadingPreviousChildPlacement() throws Exception {
+        Element<Child> child1 = testChild();
+        DocumentReference child1Application = testDocument();
+        Element<Placement> child1Placement = element(testPlacement(child1, child1Application));
+
+        CaseDetails caseDetails = CaseDetails.builder()
+            .data(Map.of(
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: Could build this in a private method?

**Comment by toireasam** [comment]
Few minor comments otherwise LGTM

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        super(uiBaseUrl, dateFormatterService, hearingBookingService);
+        this.mapper = mapper;
+    }
+
+    public Map<String, Object> buildPlacementApplicationNotificationParameters(CaseDetails caseDetails) {
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        return ImmutableMap.of(
+            "respondentLastName", getFirstRespondentLastName(caseData.getRespondents1()),
+            "caseUrl", String.format("%1$s/case/%2$s/%3$s/%4$s",
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java]: if you pull from master right now, there's a new method in EmailNotificationHelper called formatCaseUrl. Let's use that here, bit nicer

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java (hunk 1 above)
Comment (Jamesnelson117) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: whoops, good spot.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java (hunk 2 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: Discussed

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java (hunk 3 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: Discussed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java (hunk 4 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java]: the email template expects `CaseURL` but we are sending `caseUrl`. Could you confirm if its case-insensitive?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 5]
```diff
@@ -81,12 +88,18 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
         UUID childId = getSelectedChildId(caseDetails, caseData);
         Element<Child> child = placementService.getChild(caseData, childId);
 
-        Placement placement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
+        Placement currentPlacement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
             .setChild(child);
 
-        caseProperties.put("placements", placementService.setPlacement(caseData, placement));
+        Placement previousPlacement = placementService.getPlacement(caseData, child);
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: should we prefer `callbackRequest.getCaseDetailsBefore()` for getting previous `CaseData->Placement` to `placementService.getPlacement(caseData, child)`?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 6]
```diff
@@ -111,4 +124,13 @@ private void removeTemporaryFields(CaseDetails caseDetails, String... fields) {
             caseDetails.getData().remove(field);
         }
     }
+
+    private void publishPlacementApplicationUploadEvent(CallbackRequest callbackRequest, RequestData requestData) {
+        applicationEventPublisher.publishEvent(
+            new PlacementApplicationEvent(callbackRequest, requestData.authorisation(), requestData.userId()));
+    }
+
+    private Boolean isUpdatingExistingPlacement(Placement previousPlacement, Placement newPlacement) {
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: any reason why we are preferring `Boolean` and not `boolean`?

Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: Good spot, typo will change to 'boolean'

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java (hunk 5 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: I'd considered this but it gives little benefit in my opinion. At the point of call caseDetailsBefore and caseDetails will return the same data.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 7]
```diff
@@ -111,4 +124,14 @@ private void removeTemporaryFields(CaseDetails caseDetails, String... fields) {
             caseDetails.getData().remove(field);
         }
     }
+
+    private void publishPlacementApplicationUploadEvent(CallbackRequest callbackRequest, RequestData requestData) {
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: do we need to pass  `RequestData requestData` here? As its already injected above and this is a method within this class

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 8]
```diff
@@ -111,4 +124,14 @@ private void removeTemporaryFields(CaseDetails caseDetails, String... fields) {
...
     }
+
+    private void publishPlacementApplicationUploadEvent(CallbackRequest callbackRequest) {
+        applicationEventPublisher.publishEvent(
+            new PlacementApplicationEvent(callbackRequest, requestData.authorisation(), requestData.userId()));
+    }
+
+    private boolean isUpdatingExistingPlacementFilename(Placement previousPlacement, Placement newPlacement) {
+        return isNotEmpty(previousPlacement) && newPlacement.application.getFilename()
+            .equals(previousPlacement.application.getFilename());
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: just curious, do we need to verify the contents of the file also?

Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: We may in the future need to verify the document's contents but not likely to happen in FPLA-821. Document reference would be unreliable as DM will regenerate new reference ID each file upload.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 9]
```diff
@@ -92,6 +100,26 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
             .build();
     }
 
+    @PostMapping("/submitted")
+    public void handleSubmittedEvent(
+        @RequestBody @NotNull CallbackRequest callbackRequest) {
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: can `CallbackRequest` be null here?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementSubmittedControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    private Map<String, Object> buildNotificationData() {
+        return ImmutableMap.of(
+            "caseLocalAuthority", LOCAL_AUTHORITY_CODE,
+            "respondents1", List.of(
+                ImmutableMap.of(
+                    "value", Respondent.builder()
+                        .party(RespondentParty.builder()
+                            .lastName(RESPONDENT_SURNAME)
+                            .build())
+                        .build()))
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementSubmittedControllerTest.java]: any value just having 
```java
wrapElements(Respondent.builder()
                .party(RespondentParty.builder()
                    .lastName(RESPONDENT_SURNAME)
                    .build()))
```
in place of 
```java
List.of(
                ImmutableMap.of(
                    "value", Respondent.builder()
                        .party(RespondentParty.builder()
                            .lastName(RESPONDENT_SURNAME)
                            .build())
                        .build()))
```
?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        super(uiBaseUrl, dateFormatterService, hearingBookingService);
+        this.mapper = mapper;
+    }
+
+    public Map<String, Object> buildPlacementApplicationNotificationParameters(CaseDetails caseDetails) {
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        return ImmutableMap.of(
+            "respondentLastName", getFirstRespondentLastName(caseData.getRespondents1()),
+            "caseUrl", formatCaseUrl(uiBaseUrl, caseDetails.getId())
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java]: The [email template](https://www.notifications.service.gov.uk/services/12f756df-f01d-4a32-a405-e1ea8a494fbb/templates/ac0a0091-0b2e-44b9-afa1-321590b5d984) accepts `CaseUrl` but we are sending `caseUrl`. Is this case-insensitive?

Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/PlacementApplicationContentProvider.java]: It must be as it's pulling it through, reverting template to use `caseUrl` for consistency with other templates regardless.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementSubmittedControllerTest.java (hunk 10 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementSubmittedControllerTest.java]: It makes it more readable in my opinion so yes it does add value.

Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementSubmittedControllerTest.java]: Not particularly, i'll keep as it is if you don't mind.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java (hunk 9 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: Good spot, removing, can also be formatted better.

**Pull request merged**